### PR TITLE
feat(signals): add deepComputed function

### LINKED ISSUE #4202: @ngrx/signals: add `deepComputed` function
### Which @ngrx/* package(s) are relevant/related to the feature request?

signals

### Information

I would like to have a ``computed()`` function similar to the one provided by Angular for ``Signals`` but in this case, instead of returning a Signal, it would return a ``DeepSignal``.

This would allow to optimise template rendering for computed states based on objects.

Exemples:

- Current way of using computed variables:
```html
<p>
    {{ computedState().firstName }}
</p>
```
- Proposed way of using computed variables:
```html
<p>
    {{ computedState.firstName() }}
</p>
```

### Describe any alternatives/workarounds you're currently using

I'm using the ``computed()`` function provided by Angular which returns a Signal.

### I would be willing to submit a PR to fix this issue

- [ ] Yes
- [X] No

### PR DESCRIPTION
**Pull request by markostanimirovic**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

Closes #4202

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by timdeschryver — APPROVED** [review]
👍

**Comment by brandonroberts** [comment]
Q: Why wouldn't `signalState` use this?

**Comment by markostanimirovic** [comment]
> Q: Why wouldn't `signalState` use this?

`signalState` provides a state source to the `toDeepSignal` function while `deepComputed` provides a computed signal:

```ts
// signalState
const stateSource = signal(initialState);
toDeepSignal(stateSource.asReadonly());

// deepComputed
toDeepSignal(computed(computation));
```

Instead of `stateSource.asReadonly()` we can do the following:

```ts
// signalState
deepComputed(computed(() => stateSource()));
```

But I don't see a benefit because the writable signal provides the `asReadonly` method.

File: modules/signals/src/deep-computed.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import { computed } from '@angular/core';
+import { DeepSignal, toDeepSignal } from './deep-signal';
+
+export function deepComputed<T extends object>(
```
Comment (matheo) [inline modules/signals/src/deep-computed.ts:4]: @markostanimirovic what's the reasoning to restrict `T` to be an object if `toDeepSignal` supports "any"?
sometimes we have selectors that returns primitives or objects depending on the context :thinking:

Comment (markostanimirovic) [reply modules/signals/src/deep-computed.ts:4]: Yes, that's a requirement. `deepComputed` uses the initial value to create a proxy for deeply nested signals.

Comment (matheo) [reply modules/signals/src/deep-computed.ts:4]: @markostanimirovic it's basically an alias of `toDeepSignal` which doesn't have that requirement

Comment (markostanimirovic) [reply modules/signals/src/deep-computed.ts:4]: If a signal with a primitive value is provided to the `toDeepSignal` function, it's just returned without further processing. Check the implementation: https://github.com/ngrx/platform/blob/main/modules/signals/src/deep-signal.ts#L26-L29

If a primitive type or union that contains primitive is passed to the `DeepSignal` type, it will be resolved to `Signal`:

```ts
import { DeepSignal } from '@ngrx/signals';

declare const s1: DeepSignal<string>; // type: Signal<string>
declare const s2: DeepSignal<{ foo: string } | string>; // type: Signal<{ foo: string } | string>
```

I could remove the `extends object` restriction from the `deepComputed` input, but then, users may be confused why the return type is resolved to `Signal`:

```ts
const s1 = deepComputed(() => 1 as number | { foo: string }); // type: Signal<number | { foo: string }>
```

In other words, there is no difference in this example whether you use `computed` or `deepComputed`.

With the `extends object` restriction it's clear - use `deepComputed` for objects, otherwise, it has no impact and you can go with `computed`.
